File: modules/emails/src/main/java/org/noelware/charted/modules/email/DefaultEmailService.java
/*
 * 📦 charted-server: Free, open source, and reliable Helm Chart registry made in Kotlin.
 * Copyright (c) 2022-2023 Noelware, LLC. <team@noelware.org>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.noelware.charted.modules.email;

import jakarta.mail.*;
import jakarta.mail.internet.InternetAddress;
import jakarta.mail.internet.MimeMessage;
import java.util.Date;
import java.util.Properties;
import org.noelware.charted.configuration.kotlin.dsl.SMTPConfig;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class DefaultEmailService implements EmailService {
    private final Logger LOG = LoggerFactory.getLogger(DefaultEmailService.class);
    private final SMTPConfig config;

    public DefaultEmailService(SMTPConfig config) {
        this.config = config;
    }

    /**
     * Sends a test email to the {@code address} specified and any {@code content} to send.
     * @param address The email address to send to
     * @param content The content to send.
     */
    @Override
    public void sendTestEmail(String address, String content) throws MessagingException {
        final Session session = createSession();

        LOG.info("Sending email with content [{}] to address [{}]", content, address);
        final Message message = new MimeMessage(session);
        message.setFrom(new InternetAddress(config.getFrom()));
        message.setRecipient(Message.RecipientType.TO, new InternetAddress(address));
        message.setSubject("This is a test email from charted-server!");
        message.setSentDate(new Date());
        message.setText(content);

        Transport.send(message, message.getAllRecipients());
    }

    private Session createSession() {
        LOG.info("Creating SMTP session...");

        final Properties props = new Properties();
        props.put("mail.smtp.host", config.getHost());
        props.put("mail.smtp.port", config.getPort());

        if (config.getTls()) {
            props.put("mail.smtp.starttls.enable", "true");
        }

        if (config.getSsl()) {
            props.put("mail.smtp.ssl.trust", config.getHost());
        }

        if (config.getUsername() != null) {
            if (config.getPassword() == null) {
                throw new IllegalStateException("Missing `smtp.password` configuration key with `smtp.username`");
            }

            props.put("mail.smtp.auth", true);
        }

        return Session.getInstance(
                props,
                config.getUsername() == null
                        ? null
                        : new Authenticator() {
                            @Override
                            protected PasswordAuthentication getPasswordAuthentication() {
                                return new PasswordAuthentication(config.getUsername(), config.getPassword());
                            }
                        });
    }
}

File: lib/email/src/main/kotlin/org/noelware/charted/email/DefaultEmailService.kt
/*
 * 📦 charted-server: Free, open source, and reliable Helm Chart registry made in Kotlin.
 * Copyright 2022 Noelware <team@noelware.org>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.noelware.charted.email

import dev.floofy.utils.slf4j.logging
import org.noelware.charted.common.SetOnceGetValue
import org.noelware.charted.common.data.MailConfig
import java.util.Properties
import javax.mail.Authenticator
import javax.mail.Message
import javax.mail.PasswordAuthentication
import javax.mail.Session
import javax.mail.Transport
import javax.mail.internet.InternetAddress
import javax.mail.internet.MimeMessage

class DefaultEmailService(private val config: MailConfig): EmailService {
    private val _session: SetOnceGetValue<Session> = SetOnceGetValue()
    private val log by logging<DefaultEmailService>()

    init {
        log.info("Initializing session...")

        val auth = config.username != null && config.password != null
        val props = Properties().apply {
            put("mail.smtp.host", config.host)
            put("mail.smtp.port", config.port)
            put("mail.smtp.auth", auth)

            if (config.tls) put("mail.smtp.starttls.enable", "true")
            if (config.ssl) put("mail.smtp.ssl.trust", config.host)
        }

        _session.value = Session.getInstance(
            props,
            object: Authenticator() {
                override fun getPasswordAuthentication(): PasswordAuthentication? = if (auth) PasswordAuthentication(config.username!!, config.password!!) else null
            }
        )
    }

    override fun sendEmail(recipient: String, subject: String, content: String) {
        val message = MimeMessage(_session.value)
        message.setFrom(InternetAddress(config.from))
        message.addRecipient(Message.RecipientType.TO, InternetAddress(recipient))

        message.subject = subject
        message.setText(content)

        Transport.send(message)
    }
}
